### PR TITLE
fix(search-indexer): Add circularity fail-safe to error filtering

### PR DESCRIPTION
**Pull request by RunarVestmann**
# Add circularity fail-safe to error filtering

## What

* The error filtering function used to prune large fields in the error object runs forever if there's a circularity in the error object
* This change addresses that issue by keeping track of what we've visited so far and skipping depth if we've already visited a particular node

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/content-search-toolkit/src/services/elastic.service.ts [hunk 1]
```diff
@@ -109,19 +109,28 @@ export class ElasticService {
...
+      let deleted = false
+      if (Object.keys(o).length == 0) return false
+      for (const key in o) {
+        const value = o[key]
+        // Only HUMONGOUS documents should reach this limit
+        if (typeof value == 'string' && value.length > 10000) {
+          delete o[key]
+          deleted = true
+        }
+        return this.filterDoc(o[key])
```
Comment (eirikurn) [inline libs/content-search-toolkit/src/services/elastic.service.ts:129]: Should call filterDocRecursive, right?

Comment (RunarVestmann) [reply libs/content-search-toolkit/src/services/elastic.service.ts:129]: Right, you are. I'll change this

File: libs/content-search-toolkit/src/services/elastic.service.ts [hunk 2]
```diff
@@ -109,19 +109,28 @@ export class ElasticService {
...
-    for (const key in o) {
-      const value = o[key]
-      // Only HUMONGOUS documents should reach this limit
-      if (typeof value == 'string' && value.length > 10000) {
-        delete o[key]
-        deleted = true
+  private filterDoc<T>(obj: T): boolean {
+    const visited = new Set()
+    const filterDocRecursive = <O>(o: O): boolean => {
+      if (visited.has(o) || !visited) return false
```
Comment (kksteini) [inline libs/content-search-toolkit/src/services/elastic.service.ts:115]: Why the `!visited` check?

File: libs/content-search-toolkit/src/services/elastic.service.ts [hunk 3]
```diff
@@ -109,19 +109,28 @@ export class ElasticService {
...
-    if (Object.keys(o).length == 0) return false
-    for (const key in o) {
-      const value = o[key]
-      // Only HUMONGOUS documents should reach this limit
-      if (typeof value == 'string' && value.length > 10000) {
-        delete o[key]
-        deleted = true
+  private filterDoc<T>(obj: T): boolean {
+    const visited = new Set()
+    const filterDocRecursive = <O>(o: O): boolean => {
```
Comment (kksteini) [inline libs/content-search-toolkit/src/services/elastic.service.ts:114]: For better readability, is it possible to change the name `o` from a single letter?
Maybe something like `node`, or `leaf`?

**Review by kksteini — APPROVED** [review]

File: libs/content-search-toolkit/src/services/elastic.service.ts (hunk 3 above)
Comment (kristofer-lp8f) [reply libs/content-search-toolkit/src/services/elastic.service.ts:114]: :sweat_smile: Sorry that was the mathematician in me. Should definitely be more legible than that.

Comment (RunarVestmann) [reply libs/content-search-toolkit/src/services/elastic.service.ts]: I renamed it to node

Comment (kksteini) [reply libs/content-search-toolkit/src/services/elastic.service.ts]: Nice :+1:

File: libs/content-search-toolkit/README.md [hunk 4]
```diff
@@ -2,3 +2,9 @@
 
 Contains useful functions to help with querying data from content search elasticsearch.
 This library should only contain generic code
+
+## Unit testing util functions
+
+Run the following command to unit test the util functions:
+
+`yarn nx test content-search-toolkit`
```
Comment (kristofer-lp8f) [inline libs/content-search-toolkit/README.md:10]:
``````suggestion
```
yarn test content-search-toolkit
```
``````

File: libs/content-search-toolkit/src/services/utils/utils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  it('should prune fields that are longer than the letter limit', () => {
+    const test: Record<string, unknown> = {
+      a: 'some-value',
+      b: {
+        c: 'some-really-long-value-that-should-get-pruned',
+      },
+    }
+
+    // Also make the test object circular
+    test.d = test
```
Comment (kristofer-lp8f) [inline libs/content-search-toolkit/src/services/utils/utils.spec.ts:13]: Circularity should be another, separate, test case

**Review by kristofer-lp8f — APPROVED** [review]
LGTM

Only nitpicks left :smile:

File: libs/content-search-toolkit/src/services/utils/utils.spec.ts (hunk 5 above)
Comment (RunarVestmann) [reply libs/content-search-toolkit/src/services/utils/utils.spec.ts]: Agreed, I made another, very simple test for that

File: libs/content-search-toolkit/README.md (hunk 4 above)
Comment (RunarVestmann) [reply libs/content-search-toolkit/README.md]: Done

**Review by siggifv — APPROVED** [review]

**Pull request merged**